### PR TITLE
[rush-lib] components package.json JSON format validation & version adherence to semver validation

### PR DESCRIPTION
**Pull request by 1000RR**
## Summary
This PR seeks to solve 2 issues:
- rush repo components' package.json version not in semver format fails when listed as a dependency with `workspace:*` in another project - https://github.com/microsoft/rushstack/issues/2678
- when running `rush update` and there are parsing errors in a component's package.json, pnpm does not provide stderr/stdio feedback - only exit code 1, which is not descriptive to the user. The `rush update` error message is `ERROR: Error: The command failed with exit code 1`

## Details
- The component version is manually checked for being in semver format after `package.json` is loaded.
- Package.json parsing of rush repos' subprojects' validates strict JSON format - no comments, syntactically correct.

## How it was tested
- building a project containing 2 components A and B, where A imports B's `workspace:*` version. 2 tests cases:
  - B's package.json version is set to `1.0` (new error message shown, graceful exit)
  - B's package.json version is set to `1.0.0`(runs ok, graceful exit)
- building a project where a component's `package.json` is malformed JSON - with comments or extra/missing commas.

    
- [x] 👉 STEP 7: Don't forget to run "rush change": https://rushjs.io/pages/best_practices/change_logs/

### CODE REVIEW

File: libraries/node-core-library/src/JsonFile.ts [hunk 1]
```diff
@@ -113,10 +113,14 @@ export class JsonFile {
   /**
    * Loads a JSON file.
    */
-  public static load(jsonFilename: string): JsonObject {
+  public static load(jsonFilename: string, validateStrictJsonFormat?: boolean): JsonObject {
```
Comment (1000RR) [inline libraries/node-core-library/src/JsonFile.ts:116]: Only the `load` function was modified to show the package.json parsing validation change. If async and callback variants need to be changed as well, please comment.

File: libraries/node-core-library/src/JsonFile.ts [hunk 2]
```diff
@@ -113,10 +113,14 @@ export class JsonFile {
...
     try {
       const contents: string = FileSystem.readFile(jsonFilename);
-      return jju.parse(contents);
+      const parserOptions: jju.ParseOptions = {};
+      if (validateStrictJsonFormat) {
+        parserOptions.mode = 'json';
```
Comment (1000RR) [inline libraries/node-core-library/src/JsonFile.ts:121]: see https://www.npmjs.com/package/jju#jjuparse-function

File: libraries/rush-lib/src/api/RushConfigurationProject.ts [hunk 3]
```diff
@@ -106,7 +106,7 @@ export class RushConfigurationProject {
     const packageJsonFilename: string = path.join(this._projectFolder, FileConstants.PackageJson);
 
     try {
-      this._packageJson = JsonFile.load(packageJsonFilename);
+      this._packageJson = JsonFile.load(packageJsonFilename, true);
```
Comment (1000RR) [inline libraries/rush-lib/src/api/RushConfigurationProject.ts:109]: Enable JSON validation during repo's components' `package.json` parsing.

File: libraries/rush-lib/src/api/RushConfigurationProject.ts [hunk 4]
```diff
@@ -149,6 +149,13 @@ export class RushConfigurationProject {
       );
     }
 
+    if (!semver.valid(this._packageJson.version)) {
```
Comment (1000RR) [inline libraries/rush-lib/src/api/RushConfigurationProject.ts:152]: Ensure that `package.json`'s `version` field is in semver format.

**Comment by octogonz** [comment]
> * when running `rush update` and there are parsing errors in a component's package.json, pnpm does not provide stderr/stdio feedback - only exit code 1, which is not descriptive to the user. The `rush update` error message is `ERROR: Error: The command failed with exit code 1`

🥳 yes, let's finally fix this! 🙏

File: libraries/rush-lib/src/api/RushConfigurationProject.ts (hunk 4 above)
Comment (octogonz) [reply libraries/rush-lib/src/api/RushConfigurationProject.ts:152]: @elliot-nelson @D4N14L could you test this in your largest monorepos to make sure there aren't any edge cases we're not considering?

File: libraries/rush-lib/src/api/RushConfigurationProject.ts (hunk 3 above)
Comment (octogonz) [reply libraries/rush-lib/src/api/RushConfigurationProject.ts:109]: We normally would introduce an `options` parameter, since the meaning of `, true` is unclear to readers, and we may want to add additional optional parameters in the future.  I can fix this.

**Comment by octogonz** [comment]
Your new API can also fix this issue:  https://github.com/microsoft/rushstack/issues/996

File: libraries/rush-lib/src/api/RushConfigurationProject.ts (hunk 3 above)
Comment (octogonz) [reply libraries/rush-lib/src/api/RushConfigurationProject.ts]: @1000RR I pushed an update to your branch -- let me know what you think

File: libraries/node-core-library/src/JsonFile.ts [hunk 5]
```diff
@@ -39,7 +39,92 @@ export type JsonObject = any;
...
+ * Options for {@link JsonFile.parseString}, {@link JsonFile.load}, and {@link JsonFile.loadAsync}.
+ *
+ * @public
+ */
+export interface IJsonFileParseOptions {
+  /**
+   * Specifies the variant of JSON syntax to be used.
+   *
+   * @defaultValue
+   * `JsonSyntax.Json5`
```
Comment (octogonz) [inline libraries/node-core-library/src/JsonFile.ts:112]: @elliot-nelson @iclanton @D4N14L @dmichon-msft 

This is a longstanding oversight in the `JsonFile` API:

- `JsonFile.save()` always writes strict-JSON or JSON-with-comments (if we are updating a file that already contains comments)
- `JsonFile.load()` will silently accept full JSON5 syntax, for example `{ key: 0x1234, }`  (unquoted key, hex literal, trailing comma)

In practice, we don't actually use JSON5 syntax in our config files, because `JsonFile.save()` would reset it, and because every syntax highlighter would display it as an error.  In my opinion, accepting it was more of a bug than an undocumented feature. 😅

I feel strongly that if config files want to use JSON5, then they must use the `.json5` file extension.

(A similar case can be made that JSON files with comments should use a file extension such as `.jsonc` or `.cjson`, however that argument is weaker because (1) nobody seems to agree about what this file extension should be and (2) for web projects, most `.json` files are human-authored and it's ***plainly misguided*** to forbid comments in human-authored files.)

### Proposal

In a separate PR, I propose to change the default for `IJsonFileParseOptions.jsonSyntax` to be `JsonSyntax.JsonWithComments`.  It would be a MAJOR version bump for `node-core-library`, however my suspicion is that it's unlikely to break any real-world projects using this library.

Thoughts?

File: libraries/rush-lib/src/api/RushConfigurationProject.ts (hunk 4 above)
Comment (elliot-nelson) [reply libraries/rush-lib/src/api/RushConfigurationProject.ts:152]: Tested the new semver error and the "print rush update" warning, both look good.  Also tried some commands like "rush deploy" using 1000RR's branch and everything worked as normal. 👍

File: libraries/rush-lib/src/api/RushConfigurationProject.ts (hunk 3 above)
Comment (1000RR) [reply libraries/rush-lib/src/api/RushConfigurationProject.ts]: I think this works.  Please see my other note in a lower thread.

File: libraries/node-core-library/src/JsonFile.ts (hunk 5 above)
Comment (1000RR) [reply libraries/node-core-library/src/JsonFile.ts:112]: The following thoughts are likely out of scope of this bug, but discuss consistency issues amongst files with the `.json` extension in a given RushJS repo.

The disparity between the `rush.json` file being written in `cjson` syntax and every other `package.json` necessarily being written in strict `json` mode (largely due to the package managers' (at least `pnpm`'s) inability to parse plain `cjson`/`json5`) is a bit confusing in that it sets inconsistent format expectations for files with the same extension across different `.json` files in a given repo.

Anecdotally, I have observed our devs at Adobe add `cjson` syntax to `rush.json`. I imagine that it's not far-fetched to say that other projects may have made `json5` modifications to their respective `rush.json` since its parsing has been open `json5` by default. At this stage restricting `rush.json` to `cjson` would fail parsing in certain repos that have taken liberties due to suddenly stricter parsing rules of what may have effectively become a JSON5-formatted file. 

One potential avenue is to rename the `rush.json` file currently written in `cjson` format in the RushStack repo to `.json5` extension, explicitly opening the door to JSON5 format. I imagine there would be a deprecation warning when building a repo using the the older `rush.json` file, and a loading precedence or error necessary in case both a `rush.json` and `rush.json5` are found in a project.

Caveat: my suggestion would have been to stick with `cjson` format as opposed to `json5` for `rush.json` if a respective disparate file extension could be agreed upon for that file. In that case, an error for any repos with `json5` modifications to their `rush.json` would be shown.

Just my 2c.

Comment (octogonz) [reply libraries/node-core-library/src/JsonFile.ts:112]: > Anecdotally, I have observed our devs at Adobe add single line comment syntax to `rush.json`.

This is valid `cjson` and heavily used in Rush's template problems.

> I imagine that it's not far-fetched to say that other projects may have made `json5` modifications to their respective `rush.json` since its parsing has been open `json5` by default. At this stage restricting `rush.json` to `cjson` would fail parsing in certain repos that have taken liberties due to suddenly stricter parsing rules of what may have effectively become a JSON5-formatted file.

It's not far-fetched, but my hypothesis is that it is very rare.  You can test it easily by modifying your branch to change this line to say `cjson`, then running `rush install` [using your dev build](https://rushjs.io/pages/contributing/#testing-rush-builds):

https://github.com/microsoft/rushstack/blob/8a497e11561d03b44a73675c50318f37c4e0db7c/libraries/node-core-library/src/JsonFile.ts#L590-L595

I tried it with several if our repos and there were no problems.  It makes sense, because nothing about the syntax highlighting or existing file contents would suggest that JSON5 syntaxes are allowable.  If there are JSON5 syntaxes in a few places, fixing them should be pretty easy.

Also keep in mind that Rush always parses config files using the engine version specified by `rushVersion` from **rush.json**.  Thus, upgrading to the stricter Rush would be an explicit action, and thereafter there won't be any concern about users invoking the old version of Rush.

File: libraries/rush-lib/src/api/RushConfigurationProject.ts (hunk 3 above)
Comment (octogonz) [reply libraries/rush-lib/src/api/RushConfigurationProject.ts]: We need your to sign the CLA in order to get this PR merged.

**Comment by 1000RR** [comment]
Legal sign-off is pending.

**Comment by 1000RR** [comment]
@octogonz Legal has cleared the CLA - good to go.

**Comment by octogonz** [comment]
@1000RR if you're good with these changes, could you promote this to a non-draft PR?

**Comment by 1000RR** [comment]
@octogonz Looks like it's ready to go.

**Comment by octogonz** [comment]
> @octogonz may I ask for a final review and merge?

Thanks for the reminder. And thanks for your patience -- things have been very busy lately. 😄

**Review by octogonz — APPROVED** [review]
